### PR TITLE
CATROID-1099 REFACTOR ProjectCopyTask to Kotlin

### PR DESCRIPTION
**Pull request by frit96**
https://jira.catrob.at/browse/CATROID-1099
Refactoring of ProjectCopyTask to Kotlin

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catroid/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [x] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [x] Stick to the project’s gitflow workflow
- [x] Verify that your changes do not have any conflicts with the base branch
- [x] After the PR, verify that all CI checks have passed
- [ ] Post a message in the *catroid-stage* or *catroid-ide* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

**Review by adnaribo — CHANGES_REQUESTED** [review]
Please replace AsyncTask with Coroutines. More infos are in the ticket description under point 3 :)

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/ProjectListFragment.java [hunk 1]
```diff
@@ -79,7 +79,7 @@
 
 public class ProjectListFragment extends RecyclerViewFragment<ProjectData> implements
 		ProjectLoadTask.ProjectLoadListener,
-		ProjectCopyTask.ProjectCopyListener {
+		ProjectCopyTask.ProjectCopyListener{
```
Comment (hofcle78) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/ProjectListFragment.java:82]:
```suggestion
{
```

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/ProjectListFragment.java [hunk 2]
```diff
@@ -277,9 +277,7 @@ protected void copyItems(List<ProjectData> selectedItems) {
 		for (ProjectData projectData : selectedItems) {
 			String name = uniqueNameProvider.getUniqueNameInNameables(projectData.getName(), usedProjectNames);
 			usedProjectNames.add(new ProjectData(name, null, 0, false));
-			new ProjectCopyTask(projectData.getDirectory(), name)
-					.setListener(this)
-					.execute();
+			new ProjectCopyTask(projectData.getDirectory(), name, this).taskAsync();
```
Comment (hofcle78) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/ProjectListFragment.java:280]:
```suggestion
			ProjectCopyTask projectCopyTask = new ProjectCopyTask(projectData.getDirectory(), name);
			projectCopyTask.taskAsync(success -> {
				onCopyProjectComplete(success);
				return Unit.INSTANCE;
			});
```

File: catroid/src/main/java/org/catrobat/catroid/io/asynctask/ProjectCopyTask.kt [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+                destinationName
+            )
+            ProjectManager.getInstance().addNewDownloadedProject(destinationName)
+            true
+        } catch (e: IOException) {
+            Log.e(
+                TAG,
+                "Something went wrong while copying " + sourceDir.absolutePath + " to " + destinationName,
+                e
+            )
```
Comment (hofcle78) [inline catroid/src/main/java/org/catrobat/catroid/io/asynctask/ProjectCopyTask.kt:89]:
```suggestion
            Log.e(TAG, "Something went wrong while copying " + sourceDir.absolutePath + " to " + destinationName, e)
```

File: catroid/src/main/java/org/catrobat/catroid/io/asynctask/ProjectCopyTask.kt [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    private fun taskComplete(success: Boolean) {
+        weakListenerReference ?: return
+        val listener = weakListenerReference?.get()
+        listener?.onCopyFinished(success)
+    }
+
+    interface ProjectCopyListener {
+        fun onCopyFinished(success: Boolean)
+    }
+
```
Comment (hofcle78) [inline catroid/src/main/java/org/catrobat/catroid/io/asynctask/ProjectCopyTask.kt:70]:
```suggestion
```

File: catroid/src/main/java/org/catrobat/catroid/io/asynctask/ProjectCopyTask.kt [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+import java.io.File
+import java.io.IOException
+import java.lang.ref.WeakReference
+
+class ProjectCopyTask(
+    private val sourceDir: File,
+    private val destinationName: String,
+    private val listener: ProjectCopyListener
+) {
+    private var weakListenerReference: WeakReference<ProjectCopyListener>? = null
```
Comment (hofcle78) [inline catroid/src/main/java/org/catrobat/catroid/io/asynctask/ProjectCopyTask.kt:45]:
```suggestion
    private val destinationName: String
) {
```

File: catroid/src/main/java/org/catrobat/catroid/io/asynctask/ProjectCopyTask.kt [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+    @JvmOverloads
+    fun taskAsync(
+        scope: CoroutineScope = CoroutineScope(Dispatchers.IO),
```
Comment (hofcle78) [inline catroid/src/main/java/org/catrobat/catroid/io/asynctask/ProjectCopyTask.kt:49]:
```suggestion
        onCopyProjectComplete: (Boolean) -> Unit = {},
        scope: CoroutineScope = CoroutineScope(Dispatchers.IO),
```

File: catroid/src/main/java/org/catrobat/catroid/io/asynctask/ProjectCopyTask.kt [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        dispatcherOnFinished: CoroutineDispatcher = Dispatchers.Main
+    ) {
+        scope.launch {
+            weakListenerReference = WeakReference(listener)
+            val success = task(sourceDir, destinationName)
+            withContext(dispatcherOnFinished) {
+                taskComplete(success)
```
Comment (hofcle78) [inline catroid/src/main/java/org/catrobat/catroid/io/asynctask/ProjectCopyTask.kt:56]:
```suggestion
                onCopyProjectComplete(success)
```

**Review by hofcle78 — CHANGES_REQUESTED** [review]

**Review by hofcle78 — APPROVED** [review]
Looks Good!

**Review by petbuer — CHANGES_REQUESTED** [review]
Thanks a lot!
Please also refactor the corresponding test (CopyProjectTest.java) to Kotlin.

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/fragment/ProjectCopierTest.kt [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.catrobat.catroid.uiespresso.ui.fragment
+
+import android.content.Context
+import androidx.test.InstrumentationRegistry
+import androidx.test.core.app.ApplicationProvider
```
Comment (hofcle78) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/fragment/ProjectCopierTest.kt:27]:
```suggestion
import androidx.test.platform.app.InstrumentationRegistry.getInstrumentation
import androidx.test.core.app.ApplicationProvider.getApplicationContext
```

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/fragment/ProjectCopierTest.kt [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+        createProject(toBeCopiedProjectName)
+        baseActivityTestRule.launchActivity(null)
+    }
+
+    @Category(AppUi::class, Smoke::class)
+    @Test
+    fun copyProjectTest() {
+        openActionBarOverflowOrOptionsMenu(
+            InstrumentationRegistry.getInstrumentation().targetContext
+        )
```
Comment (hofcle78) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/fragment/ProjectCopierTest.kt:84]:
```suggestion
        openActionBarOverflowOrOptionsMenu(getInstrumentation().targetContext)
```

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/fragment/ProjectCopierTest.kt [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+        onView(ViewMatchers.withText("$toBeCopiedProjectName (3)"))
+            .check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
+    }
+
+    @Throws(IOException::class)
+    private fun createProject(projectName: String) {
+        val project = Project(
+            ApplicationProvider.getApplicationContext(),
+            projectName
+        )
```
Comment (hofcle78) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/fragment/ProjectCopierTest.kt:133]:
```suggestion
        val project = Project(getApplicationContext(), projectName)
```

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/fragment/ProjectCopierTest.kt [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+        val script: Script = StartScript()
+        script.addBrick(SetXBrick(Formula(BrickValues.X_POSITION)))
+        sprite.addScript(script)
+        project.defaultScene.addSprite(sprite)
+        ProjectManager.getInstance().currentProject = project
+        ProjectManager.getInstance().currentSprite = sprite
+        XstreamSerializer.getInstance().saveProject(project)
+        val soundFile = ResourceImporter.createSoundFileFromResourcesInDirectory(
+            androidx.test.platform.app.InstrumentationRegistry.getInstrumentation().context
+                .resources,
```
Comment (hofcle78) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/fragment/ProjectCopierTest.kt:144]:
```suggestion
            getInstrumentation().context.resources,
```

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/fragment/ProjectCopierTest.kt [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+            org.catrobat.catroid.test.R.raw.longsound,
+            File(
+                project.defaultScene.directory,
+                Constants.SOUND_DIRECTORY_NAME
+            ),
```
Comment (hofcle78) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/fragment/ProjectCopierTest.kt:149]:
```suggestion
            File(project.defaultScene.directory, Constants.SOUND_DIRECTORY_NAME),
```

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/fragment/ProjectCopierTest.kt [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+        val backgroundSprite = Sprite(
+            ApplicationProvider.getApplicationContext<Context>().getString(R.string.background)
+        )
+        backgroundSprite.look.zIndex = Constants.Z_INDEX_BACKGROUND
+        secondScene.addSprite(backgroundSprite)
+        project.addScene(secondScene)
+        saveProjectSerial(
+            project,
+            ApplicationProvider.getApplicationContext()
+        )
```
Comment (hofcle78) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/fragment/ProjectCopierTest.kt:182]:
```suggestion
        saveProjectSerial(project, getApplicationContext())
```

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/fragment/ProjectCopierTest.kt [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+        val lookDataList = sprite.lookList
+        val lookData = LookData()
+        lookData.file = imageFile
+        lookData.name = "testLook1"
+        lookDataList.add(lookData)
+        val secondScene = Scene("secondScene", project)
+        val backgroundSprite = Sprite(
+            ApplicationProvider.getApplicationContext<Context>().getString(R.string.background)
+        )
```
Comment (hofcle78) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/fragment/ProjectCopierTest.kt:175]:
```suggestion
        val backgroundSprite = Sprite(getApplicationContext<Context>().getString(R.string.background))
```

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/fragment/ProjectCopierTest.kt [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+            "longsound.mp3"
+        )
+        val soundInfoList = sprite.soundList
+        val soundInfo = SoundInfo()
+        soundInfo.file = soundFile
+        soundInfo.name = "testSound1"
+        soundInfoList.add(soundInfo)
+        val imageFile = ResourceImporter.createImageFileFromResourcesInDirectory(
+            androidx.test.platform.app.InstrumentationRegistry.getInstrumentation().context
+                .resources,
```
Comment (hofcle78) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/fragment/ProjectCopierTest.kt:159]:
```suggestion
            getInstrumentation().context.resources,
```

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/fragment/ProjectCopierTest.kt [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+            org.catrobat.catroid.test.R.drawable.catroid_banzai,
+            File(
+                project.defaultScene.directory,
+                Constants.IMAGE_DIRECTORY_NAME
+            ),
+            "catroid_sunglasses.png", 1.0
```
Comment (hofcle78) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/fragment/ProjectCopierTest.kt:165]:
```suggestion
            File(project.defaultScene.directory, Constants.IMAGE_DIRECTORY_NAME),
            "catroid_sunglasses.png",
            1.0
```

**Review by hofcle78 — CHANGES_REQUESTED** [review]

**Comment by frit96** [comment]
I made all the requested changes

**Review by hofcle78 — APPROVED** [review]

**Review by petbuer — APPROVED** [review]
Thanks!

**Review by adnaribo — APPROVED** [review]

**Pull request merged**